### PR TITLE
core: replace secp256k with k256 in crypto::ecdsa

### PR DESCRIPTION
**Pull request by michalkucharczyk**
This PR replaces the usage of [secp256k](https://crates.io/crates/secp256k1) crate with [k256](https://crates.io/crates/k256) in `core::crypto::ecdsa` for `non-std` environments as outcome of discussion in #3448.

`secp256k1` is used in `std`, meaning that we should not affect host performance with this PR.
`k256` is enabled in runtimes (`no-std`), and is required to proceed with #2044.

If desirable, in future we can switch to `k256` also for `std`. That would require some performance evaluation (e.g. for EVM chains as per https://github.com/paritytech/polkadot-sdk/issues/3448#issuecomment-1976780391).

Closes https://github.com/paritytech/polkadot-sdk/issues/3448

### CODE REVIEW

**Comment by michalkucharczyk** [comment]
cc: @burdges

File: substrate/primitives/core/src/ecdsa.rs [hunk 1]
```diff
@@ -331,30 +333,19 @@ impl Signature {
...
-			.ok()
-			.map(|pubkey| Public(pubkey.serialize()))
+		VerifyingKey::recover_from_prehash(message, &sig, rid).map(|p| p.into()).ok()
 	}
 }
 
 #[cfg(feature = "full_crypto")]
-impl From<RecoverableSignature> for Signature {
-	fn from(recsig: RecoverableSignature) -> Signature {
+impl From<(k256::ecdsa::Signature, RecoveryId)> for Signature {
```
Comment (michalkucharczyk) [inline substrate/primitives/core/src/ecdsa.rs:344]: `(k256::ecdsa::Signature, RecoveryId)` is a result of `sign_prehash_recoverable`. Convenient `From`.

File: substrate/primitives/core/src/ecdsa.rs [hunk 2]
```diff
@@ -455,14 +438,10 @@ impl Pair {
...
-
-		#[cfg(feature = "std")]
-		let context = SECP256K1;
-		#[cfg(not(feature = "std"))]
-		let context = Secp256k1::signing_only();
-
-		context.sign_ecdsa_recoverable(&message, &self.secret).into()
+		self.secret
+			.sign_prehash_recoverable(message)
+			.expect("signing may not fail (???). qed.")
```
Comment (michalkucharczyk) [inline substrate/primitives/core/src/ecdsa.rs:443]: In out API (e.g. `TraitPair`) we assume signing cannot fail (return type is not Option or Result).
Is `expect` fine here? What would be condition for potential failure?

File: substrate/primitives/core/src/ecdsa.rs [hunk 3]
```diff
@@ -498,18 +477,6 @@ impl Pair {
...
-// NOTE: this solution is not effective when `Pair` is moved around memory.
-// The very same problem affects other cryptographic backends that are just using
-// `zeroize`for their secrets.
-#[cfg(feature = "full_crypto")]
-impl Drop for Pair {
-	fn drop(&mut self) {
-		self.secret.non_secure_erase()
-	}
-}
-
```
Comment (michalkucharczyk) [inline substrate/primitives/core/src/ecdsa.rs:512]: note: `Drop` shall be fine in [`k256::SigningKey`](https://docs.rs/ecdsa/0.16.9/src/ecdsa/signing.rs.html#380-384)

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5396533

File: substrate/primitives/core/src/ecdsa.rs [hunk 4]
```diff
@@ -131,6 +123,16 @@ impl AsMut<[u8]> for Public {
 	}
 }
 
+impl From<VerifyingKey> for Public {
+	fn from(pubkey: VerifyingKey) -> Self {
+		Self::unchecked_from(
+			pubkey.to_sec1_bytes()[..]
+				.try_into()
+				.expect("valid key is serializable to [u8,33]. qed."),
```
Comment (michalkucharczyk) [inline substrate/primitives/core/src/ecdsa.rs:131]: double-check needed here. Can we actually fail?

File: substrate/primitives/core/Cargo.toml [hunk 5]
```diff
@@ -52,9 +52,9 @@ blake2 = { version = "0.10.4", default-features = false, optional = true }
 libsecp256k1 = { version = "0.7", default-features = false, features = ["static-context"], optional = true }
 schnorrkel = { version = "0.11.4", features = ["preaudit_deprecated"], default-features = false }
 merlin = { version = "3.0", default-features = false }
-secp256k1 = { version = "0.28.0", default-features = false, features = ["alloc", "recovery"], optional = true }
 sp-crypto-hashing = { path = "../crypto/hashing", default-features = false, optional = true }
 sp-runtime-interface = { path = "../runtime-interface", default-features = false }
+k256 = { version = "0.13.3", features = ["alloc", "ecdsa"], default-features = false }
```
Comment (michalkucharczyk) [inline substrate/primitives/core/Cargo.toml:57]: default features are:
[`["arithmetic", "ecdsa", "pkcs8", "precomputed-tables", "schnorr", "std"]`](https://github.com/RustCrypto/elliptic-curves/blob/cdca0123ab36efaeaab92ef728a0912edcc6f164/k256/Cargo.toml#L44)

I think we shall be fine with just 'alloc' and 'ecdsa'.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5397556

**Comment by michalkucharczyk** [comment]
bot fmt

**Review by davxy — COMMENTED** [review]
I'll give it a try tomorrow. Maybe before merge we should try to sync from genesis (aka *burn-in* test)?

File: substrate/primitives/core/src/ecdsa.rs [hunk 6]
```diff
@@ -53,7 +47,7 @@ pub const PUBLIC_KEY_SERIALIZED_SIZE: usize = 33;
 /// The byte length of signature
 pub const SIGNATURE_SERIALIZED_SIZE: usize = 65;
 
-/// A secret seed (which is bytewise essentially equivalent to a SecretKey).
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/primitives/core/src/ecdsa.rs:56]: doc improved in: 35878b20bbb

File: substrate/primitives/core/Cargo.toml (hunk 5 above)
Comment (burdges) [reply substrate/primitives/core/Cargo.toml:57]: precomputed-tables should impact performance significantly if this gets used in the host.  also std of course.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5429611

File: substrate/primitives/core/Cargo.toml [hunk 7]
```diff
@@ -75,7 +76,15 @@ harness = false
 bench = false
 
 [features]
-default = ["std"]
+default = ["backend_secp256k1", "std"]
+
+# use secp256k1 crate, better performance, intended to be used on host side
+backend_secp256k1 = [
```
Comment (davxy) [inline substrate/primitives/core/Cargo.toml:82]:
```suggestion
ecdsa-backend-secp256k1 = [
```
Same for k256

File: substrate/primitives/core/src/ecdsa.rs [hunk 8]
```diff
@@ -331,23 +369,44 @@ impl Signature {
...
+		#[cfg(feature = "backend_secp256k1")]
+		{
+			let rid = RecoveryId::from_i32(self.0[64] as i32).ok()?;
+			let sig = RecoverableSignature::from_compact(&self.0[..64], rid).ok()?;
+			let message = Message::from_digest_slice(message).expect("Message is 32 bytes; qed");
+
+			#[cfg(feature = "std")]
+			let context = SECP256K1;
+			#[cfg(not(feature = "std"))]
+			let context = Secp256k1::verification_only();
```
Comment (davxy) [inline substrate/primitives/core/src/ecdsa.rs:381]: maybe this can be removed at this point?

**Comment by davxy** [comment]
Maybe without leaving too much choice (choices sometimes makes the user confused 😆) we can just use secp256k1 on std and k256 for no-std. I mean, without introducing these two mutually exclusive features

**Comment by michalkucharczyk** [comment]
> Maybe without leaving too much choice (choices sometimes makes the user confused 😆) we can just use secp256k1 on std and k256 for no-std. I mean, without introducing these two mutually exclusive features.
> 
> I'd personally find it easier if I have no idea what I'm doing and what is best

I am not entirely convinced if that is good approach -- need to think about it. We're going to get rid of `no-std` in favor of `substrate-runtime`. Also without `backend_xxx` feature we are unable to select single secp256k1 crate.

Anyway I can do it in #2044 (which I hope will be merged soon) - let's discuss it there, is that OK for you?

File: substrate/primitives/core/src/ecdsa.rs [hunk 9]
```diff
@@ -331,23 +362,41 @@ impl Signature {
...
-		let context = SECP256K1;
+		{
+			let rid = RecoveryId::from_i32(self.0[64] as i32).ok()?;
+			let sig = RecoverableSignature::from_compact(&self.0[..64], rid).ok()?;
+			let message = Message::from_digest_slice(message).expect("Message is 32 bytes; qed");
+
+			#[cfg(feature = "std")]
+			let context = SECP256K1;
+			#[cfg(not(feature = "std"))]
+			let context = Secp256k1::verification_only();
```
Comment (michalkucharczyk) [inline substrate/primitives/core/src/ecdsa.rs:374]:
```suggestion
```

File: substrate/primitives/core/src/ecdsa.rs [hunk 10]
```diff
@@ -331,23 +362,41 @@ impl Signature {
...
-		let message = Message::from_digest_slice(message).expect("Message is 32 bytes; qed");
-
 		#[cfg(feature = "std")]
-		let context = SECP256K1;
+		{
+			let rid = RecoveryId::from_i32(self.0[64] as i32).ok()?;
+			let sig = RecoverableSignature::from_compact(&self.0[..64], rid).ok()?;
+			let message = Message::from_digest_slice(message).expect("Message is 32 bytes; qed");
+
+			#[cfg(feature = "std")]
```
Comment (michalkucharczyk) [inline substrate/primitives/core/src/ecdsa.rs:371]:
```suggestion
```

File: substrate/primitives/core/src/ecdsa.rs [hunk 11]
```diff
@@ -384,17 +433,26 @@ impl TraitPair for Pair {
...
-		#[cfg(not(feature = "std"))]
-		let context = Secp256k1::signing_only();
+		{
+			let secret = SecretKey::from_slice(seed_slice)
+				.map_err(|_| SecretStringError::InvalidSeedLength)?;
+
+			#[cfg(feature = "std")]
+			let context = SECP256K1;
+			#[cfg(not(feature = "std"))]
+			let context = Secp256k1::signing_only();
```
Comment (michalkucharczyk) [inline substrate/primitives/core/src/ecdsa.rs:444]:
```suggestion
```

File: substrate/primitives/core/src/ecdsa.rs [hunk 12]
```diff
@@ -455,14 +520,25 @@ impl Pair {
...
-		let message = Message::from_digest_slice(message).expect("Message is 32 bytes; qed");
-
 		#[cfg(feature = "std")]
-		let context = SECP256K1;
-		#[cfg(not(feature = "std"))]
-		let context = Secp256k1::signing_only();
+		{
+			let message = Message::from_digest_slice(message).expect("Message is 32 bytes; qed");
+
+			#[cfg(feature = "std")]
```
Comment (michalkucharczyk) [inline substrate/primitives/core/src/ecdsa.rs:527]:
```suggestion
```

File: substrate/primitives/core/src/ecdsa.rs [hunk 13]
```diff
@@ -455,14 +520,25 @@ impl Pair {
...
+			let context = SECP256K1;
+			#[cfg(not(feature = "std"))]
+			let context = Secp256k1::signing_only();
```
Comment (michalkucharczyk) [inline substrate/primitives/core/src/ecdsa.rs:530]:
```suggestion
```

**Comment by michalkucharczyk** [comment]
> I'll give it a try tomorrow. Maybe before merge we should try to sync from genesis (aka _burn-in_ test)?

Sync from genesis test on kusama and polkadot was fine (with `k256` on host side).

**Review by davxy — APPROVED** [review]
LGTM

File: substrate/primitives/core/src/ecdsa.rs [hunk 14]
```diff
@@ -96,18 +97,27 @@ impl Public {
...
+		let pubkey = {
+			#[cfg(feature = "std")]
+			{
+				PublicKey::from_slice(&full)
+			}
+			#[cfg(not(feature = "std"))]
+			{
+				VerifyingKey::from_sec1_bytes(&full)
+			}
+		};
```
Comment (davxy) [inline substrate/primitives/core/src/ecdsa.rs:119]:
```suggestion
		#[cfg(feature = "std")]
		let pubkey = PublicKey::from_slice(&full);
		#[cfg(not(feature = "std"))]
		let pubkey = VerifyingKey::from_sec1_bytes(&full);
```
This is just personal preference. Feel free to ignore :-D

File: substrate/primitives/core/src/ecdsa.rs [hunk 15]
```diff
@@ -438,7 +479,14 @@ impl TraitPair for Pair {
...
 	pub fn seed(&self) -> Seed {
-		self.secret.secret_bytes()
+		#[cfg(feature = "std")]
+		{
+			self.secret.secret_bytes()
+		}
+		#[cfg(not(feature = "std"))]
+		{
+			self.secret.to_bytes().into()
+		}
```
Comment (davxy) [inline substrate/primitives/core/src/ecdsa.rs:489]:
```suggestion
		#[cfg(feature = "std")]
		self.secret.secret_bytes()
		#[cfg(not(feature = "std"))]
		self.secret.to_bytes().into()
```

Comment (michalkucharczyk) [reply substrate/primitives/core/src/ecdsa.rs:489]: Unfortunately we need to keey `{...}` here, as what you proposed does not compile:
```
error: expected `;`, found `#`
   --> substrate/primitives/core/src/ecdsa.rs:483:30
    |
483 |             self.secret.secret_bytes()
    |                                       ^ help: add `;` here
484 |         #[cfg(not(feature = "std"))]
    |         - unexpected token
```

This is why I kept `{}` above (for consistency).

File: substrate/primitives/core/Cargo.toml [hunk 16]
```diff
@@ -94,6 +98,7 @@ std = [
 	"hash256-std-hasher/std",
 	"impl-serde/std",
 	"itertools",
+	"k256/std",
```
Comment (bkchr) [inline substrate/primitives/core/Cargo.toml:101]: We don't really need to enable `std`?

**Review by bkchr — APPROVED** [review]

File: substrate/primitives/core/Cargo.toml (hunk 16 above)
Comment (michalkucharczyk) [reply substrate/primitives/core/Cargo.toml:101]: thanks, leftover from prev iteration.
```suggestion
```

Comment (michalkucharczyk) [reply substrate/primitives/core/Cargo.toml]: hm..., looks like we need it, c[heck-features](https://github.com/paritytech/polkadot-sdk/actions/runs/8208792794/job/22452944449?pr=3525) is failing:
```
Run cargo featalign substrate/bin/node/runtime \
  cargo featalign substrate/bin/node/runtime \
    --features std,runtime-benchmarks,try-runtime \
    --mode check \
    --workspace-only \
    --ignore sc-executor \
    --default-std \
    --depth -1 | jq
  shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
{
  "sp-core 28.0.0 (path+file:///home/runner/work/polkadot-sdk/polkadot-sdk/substrate/primitives/core)": [
    {
      "id": "k256 0.13.3 (registry+https://github.com/rust-lang/crates.io-index)",
      "alias": "k256",
      "dependency-path": "/kitchensink-runtime/frame-benchmarking/frame-support/sp-api/sp-core",
      "problem": {
        "missing-features": [
          "std"
        ]
      }
    }
  ]
}
Error: Process completed with exit code 255
```

**Pull request merged**